Make Gophish Docker data persistent

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR makes Docker volume data on Gophish instances persistent across instance restarts and re-deployments.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

The primary driver of this change was to avoid having the postfix DKIM key change every time the PCA-Gophish composition or Gophish instance was restarted/redeployed.  To accomplish this, we attach an EBS volume to each Gophish instance and we tell Docker to use that EBS volume as the [`data-root` directory](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon).  This allows for any Docker data to be made persistent, as long as it lives on a Docker volume.

Notes:
* The existing EFS volume is not suitable for this purpose because EFS is not supported as a [backing file system for Docker](https://docs.docker.com/storage/storagedriver/select-storage-driver/#supported-backing-filesystems).
* Issues https://github.com/cisagov/pca-gophish-composition-packer/issues/48 and https://github.com/cisagov/cool-assessment-terraform/issues/120 were created to move the Gophish persistent data directory from it's current location on the EFS volume (via a bind mount) to the new EBS volume (via a named volume).
* Postfix user passwords are not persisted at this time because of how the [postfix-docker entrypoint script](https://github.com/cisagov/postfix-docker/blob/49d4ce8f030f044c8784814f934f399b3f5cea35/src/docker-entrypoint.sh#L73-L74) is written, but that could be changed in the future if desired.  Currently, the postfix user passwords are recreated (unless they have been hardcoded as Docker secrets) every time the Gophish instance or the PCA-Gophish composition or the postfix service in the PCA-Gophish is restarted.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

Closes https://github.com/cisagov/pca-gophish-composition-packer/issues/46 and is part of the work for https://github.com/cisagov/cool-system/issues/90.

## 🧪 Testing ##

Testing was performed in a Staging environment.  

I first verified that a freshly created Gophish instance:
* [x] Mounted up it's newly-created EBS volume and was storing Docker data on it
* [x] The PCA-Gophish composition successfully started automatically
* [x] The Gophish GUI was usable via Guacamole
* [x] A DKIM key was created by the postfix service

Next, I verified that the PCA-Gophish composition successfully started automatically and the previously-created DKIM key and postfix configuration persisted when I did each of the following actions:
* [x] Destroyed the Gophish instance and created a new one to replace it
* [x] Stopped the Gophish instance (via the AWS console) and started it again
* [x] Rebooted the Gophish instance (via the AWS console)
* [x] Restarted the `pca-gophish-composition` service (`systemctl restart pca-gophish-composition.service`
* [x] Restarted the PCA-Gophish composition (`cd /var/pca/pca-gophish-composition; docker-compose restart`)
* [x] Restarted the Docker service (`systemctl restart docker.service`)

In each case above, verification of the DKIM key was done via:
```console
cd /var/pca/pca-gophish-composition
docker-compose logs | grep DKIM
```

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

dav3r committed:
Add EBS volume setup script

This script is used to prepare an EBS volume for use by creating a file 
system (if necessary), mounting it, and ensuring that it will remount 
after a reboot.

dav3r committed:
Add script to copy Docker data from default data-root dir to a new dir

dav3r committed:
Add cloud-init script to write /etc/docker/daemon.json

daemon.json file is used to define a new data-root directory (override 
the default of /var/lib/docker).

dav3r committed:
Define two new locals related to persistent Docker volumes

dav3r committed:
Create and attach an EBS volume to each Gophish instance

This allows Docker volume data to persist across instance restarts and 
redeployments.

Note that Docker data cannot be stored on the existing EFS volume 
because EFS is not supported as a backing file system for Docker: 
https://docs.docker.com/storage/storagedriver/select-storage-driver/#supported-backing-filesystems

dav3r committed:
Run Gophish cloud-init script to write /etc/docker/daemon.json

In daemon.json, we specify that the new Docker data-root directory 
(where Docker stores volume data) is the directory where the Docker EBS 
volume is mounted on the Gophish instance.  This enables Docker data to 
persist on the EBS volume.

dav3r committed:
Run cloud-init scripts to set up EBS volume as the Docker data-root

dav3r committed:
Add additional EC2 permissions needed to manage EBS volumes

jsf9k commented on gophish_cloud_init.tf:
```diff
@@ -37,9 +37,55 @@ data "cloudinit_config" "gophish_cloud_init_tasks" {
     merge_type   = "list(append)+dict(recurse_array)+str()"
   }
 
+  # Prepare and mount EBS volume to hold Docker data-root data.
+  # Note that this script and the next one must take place in a certain order,
+  # so we prepend numbers to the script names to force that to that happen.
+  #
+  # Here is where the user scripts are called by cloud-init:
+  # https://github.com/canonical/cloud-init/blob/master/cloudinit/config/cc_scripts_user.py#L45
+  #
+  # And here is where you can see how cloud-init sorts the scripts:
+  # https://github.com/canonical/cloud-init/blob/master/cloudinit/subp.py#L373
```
This comment is amazing!!! :smile:

dav3r replied:
![image](https://user-images.githubusercontent.com/21343441/119693043-eb88c100-be19-11eb-9852-95ff259e5210.png)

felddy replied:
Since you stole this from Dad, you get the same comment.  Does it make sense to space these out to allow thangs in between?  `10`, `20`...  
![iu-1](https://user-images.githubusercontent.com/3229435/119889858-1996ff80-bf05-11eb-8471-4a423a4819d7.jpeg)

dav3r replied:
> Since you stole this from Dad, you get the same comment. Does it make sense to space these out to allow thangs in between? `10`, `20`...

I'll steal his reply too:  https://github.com/cisagov/cool-assessment-terraform/pull/118#discussion_r640925051

jsf9k commented on gophish_ec2.tf:
```diff
@@ -88,3 +91,30 @@ resource "aws_eip_association" "gophish" {
+# data across instance restarts and redeployments.  Note that Docker data
+# cannot be stored on the existing EFS volume because EFS is not supported
+# as a backing file system for Docker:
+# https://docs.docker.com/storage/storagedriver/select-storage-driver/#supported-backing-filesystems
```
Glad you included this link.

jsf9k reviewed (commented):
Pre-reviewed and left a few laudatory comments about comments.

dav3r committed:
Add a TODO comment to persist Gophish data on EBS volume, instead of EFS volume

For details, see 
https://github.com/cisagov/cool-assessment-terraform/issues/120

jsf9k reviewed (approved):
Nice work here!

felddy commented on cloud-init/copy-docker-data-to-new-root-dir.tpl.sh:
```diff
@@ -0,0 +1,41 @@
+#
+# shellcheck disable=SC2154
+set -o nounset
+set -o errexit
```
Why you no `pipefail`?

dav3r replied:
I didn't use any pipes, so I figured I didn't need it.  But if we want to keep it in there as a standard thang, I'm ok with that too.

jsf9k replied:
I'd prefer to keep it in, since if someone else comes along and adds pipes they'll expect it to be there.  It should be a standard thang.

dav3r replied:
You win:  https://github.com/cisagov/cool-assessment-terraform/pull/119/commits/3a3e080b5cc0a51a9731c22b2a524990e39b5924

felddy reviewed (approved):
So choice.

dav3r committed:
Add pipefail option

Even though there are no pipes in use now, that doesn't mean some won't 
be used in the future.

mcdonnnj commented on cloud-init/copy-docker-data-to-new-root-dir.tpl.sh:
```diff
@@ -0,0 +1,42 @@
+# Docker data
+
+# This is a Terraform template file, and the input variables are
+# passed in via the templatefile().
```
```suggestion
# passed in via templatefile().
```

dav3r replied:
Fixed in https://github.com/cisagov/cool-assessment-terraform/pull/119/commits/9e1cb0c4679245d9de04d8bdd9e131e4da5745a4.

mcdonnnj commented on cloud-init/ebs-disk-setup.tpl.sh:
```diff
@@ -0,0 +1,74 @@
+# mounting (defaults or defaults,noauto, for example)
+
+# This is a Terraform template file, and the input variables are
+# passed in via the templatefile().
```
```suggestion
# passed in via templatefile().
```

dav3r replied:
Fixed in https://github.com/cisagov/cool-assessment-terraform/pull/119/commits/9e1cb0c4679245d9de04d8bdd9e131e4da5745a4.

mcdonnnj commented on gophish_cloud_init.tf:
```diff
@@ -37,11 +37,60 @@ data "cloudinit_config" "gophish_cloud_init_tasks" {
+
   # Set up everything needed to successfully launch the
   # pca-gophish-composition service.
+  # TODO Persist Gophish data on EBS volume, instead of EFS volume.
```
```suggestion
  # TODO Persist GoPhish data on EBS volume, instead of EFS volume.
```

dav3r replied:
According to the [documentation](https://getgophish.com/documentation/) "Gophish" is how they spell it.  What I really need to do is change all of the other instances of "GoPhish" to "Gophish"... so that is what I did in https://github.com/cisagov/cool-assessment-terraform/pull/119/commits/57939e6083a20e9353186a43e4a3e81f7c5b0de0.

mcdonnnj reviewed (approved):
Strong work 💪💪💪 I just had a few small suggestions.

dav3r committed:
Improve grammar

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use preferred capitalization for Gophish

This is based on the Gophish docs: https://getgophish.com/documentation/
